File: algorithm/merge/sort.go
package main

import (
	"fmt"
)

// 归并排序
func Merge(slice []int, left int, right int) {

	if left < right {
		mid := (left + right) / 2
		Merge(slice, left, mid)
		Merge(slice, mid+1, right)

		var temp []int
		i := left
		j := mid + 1
		for i <= mid && j <= right {
			if slice[i] <= slice[j] {
				temp = append(temp, slice[i])
				i++
				continue
			}
			temp = append(temp, slice[j])
			j++
		}
		if i <= mid {
			temp = append(temp, slice[i:mid+1]...)
		}
		if j <= right {
			temp = append(temp, slice[j:right+1]...)
		}

		for i, v := range temp {
			slice[left+i] = v
		}
	}
}

func main() {
	slice := []int{2, 3, 1, 4, 32, 42, 88, 21, 6, 8, 9, 5, 14, 12, 18, 54, 7, 35, 36, 24, 13}
	Merge(slice, 0, len(slice)-1)
	fmt.Println(slice)
}

File: algorithm/merge/sort.py
import random


def merge(list, left, right):
    if left < right:
        mid = (left + right) // 2
        merge(list, left, mid)
        merge(list, mid + 1, right)

        tmp = []
        i = left
        j = mid + 1

        while i <= mid and j <= right:
            if list[i] <= list[j]:
                tmp.append(list[i])
                i += 1
            else:
                tmp.append(list[j])
                j += 1
        if i <= mid:
            tmp.extend(list[i:mid + 1])

        if j <= right:
            tmp.extend(list[j:right + 1])

        list[left:right + 1] = tmp


list = []
for i in range(1, 51):
    list.append(i)

random.shuffle(list)
merge(list, 0, len(list) - 1)
print(list)
